lib/routes/account.js

```diff
@@ -1021,50 +1048,78 @@ module.exports = function (
-        db.account(Buffer(uid, 'hex'))
-          .then(
-            function (account) {
-              // If the account is already verified, they may be e.g.
-              // clicking a stale link.  Silently succeed.
-              if (account.emailVerified) {
-                log.increment('account.already_verified')
-                return true
-              }
-              if (!butil.buffersAreEqual(code, account.emailCode)) {
-                throw error.invalidVerificationCode()
-              }
-              log.timing('account.verified', Date.now() - account.createdAt)
-              log.event('verified', { email: account.email, uid: account.uid, locale: account.locale })
-              log.increment('account.verified')
 
-              // send a push notification to all devices that the account changed
-              push.notifyUpdate(uid)
+        if(code.length === 32) {
+          // Hex code length of 32 signifies a tokenId, perform login token verification
+          log.begin('Account.RecoveryLoginVerify', request);
+          db.verifyToken(code, {uid: uid})
+            .then(function () {
+            })
+            .done(
+              function () {
+                reply({})
+              },
+              reply
+            )
+        } else {
+          log.begin('Account.RecoveryEmailVerify', request)
+          db.account(Buffer(uid, 'hex'))
+            .then(
+              function (account) {
+                // If the account is already verified, they may be e.g.
+                // clicking a stale link.  Silently succeed.
+                if (account.emailVerified) {
+                  log.increment('account.already_verified')
+                  return true
+                }
+                if (!butil.buffersAreEqual(code, account.emailCode)) {
+                  throw error.invalidVerificationCode()
+                }
+                log.timing('account.verified', Date.now() - account.createdAt)
+                log.event('verified', {email: account.email, uid: account.uid, locale: account.locale})
+                log.increment('account.verified')
 
-              return db.verifyEmail(account)
-                .then(mailer.sendPostVerifyEmail.bind(
-                    mailer,
-                    account.email,
-                    {
-                      acceptLanguage: request.app.acceptLanguage
-                    }
-                  )
-                )
-            }
-          )
-          .done(
-            function () {
-              reply({})
-            },
-            reply
-          )
+                // send a push notification to all devices that the account changed
+                push.notifyUpdate(uid)
+
+                return db.sessions(account.uid)
+                  .then(function(sessions) {
+                    // TODO How to know which session is associated with the email verification?
+                    // Assuming last created session?
```

Comment:
@rfk - does an emailCode need to be stored on the account? If so, what is it used for (I imagine for verification!) Can the account be determined from a sessionToken or keyFetchToken?

I can imagine the new scheme working as follows:
User signs up, no code is stored on the account. A sessionToken, and possibly a keyFetchToken are created, both unverified. A code is associated with both of those. When the user goes to verify, the code is first looked for in the sessionToken table. The returned sessionToken has a uid that can be used as a key into the account table. The account is fetched and marked as verified. 

If the code is not found in the sessionToken table, try to see if it's a legacy code. Look in the Account table for the code. If found, verify the account.

If a signup then subsequent signin of an unverified account both return distinct codes, I imagine the scheme would *have* to work that way so that the account could be marked as verified.